magento/magento2
# Adding in customisable exclude paths in setup:di:compile

### Description (*)

There has been lots of demand for extending the di compilation exclude paths (either via CLI options or events) so I decided to add in an event `setup_di_compile_excluded_patterns` that can be observed and added to from within a third-party module.

### Fixed Issues (if relevant)

1. magento/magento2#5676: [M2.1] - No CLI option to exclude patterns on setup:di:compile and other inflexibilities

### Manual testing scenarios (*)


1. Create a module that normally breaks setup:di:compile
2. Add an observer for the 'setup_di_compile_excluded_patterns' event that adds an exclusion path for the file that is breaking the compilation
3. Run `bin/magento setup:di:compile` and see it compiles as expected despite the bad code in the module file.


### Contribution checklist (*)

### Notes
There are a bunch of commits on here, but as you can tell from the diff, those proposed changes that were rejected previously have been reverted. This simply updates the setup:di:compile to give the ability to exclude custom paths via an observer.
